Increase precision for golden solution in transformer tests.

anatoliylitv opened the pull request:
# Overview

Force FP32 precision for "golden" solution computation when TF32 is set for compute test solution.

# Rationale

The `test/test_transformers.py` testing suite calculates the numerical
tolerance by comparing output tensors from the same precision ("reference")
and higher precision ("golden"), both calculated by `SDPBackend.MATH`.
However, if the golden output is calculated with TF32 rather than FP32, which in
fact is less accurate than the FA/ME backend if they used IEEE rather than
TF32 for their accumulation.

The loss of precison causes false negatives in SDPA tests like
`TestSDPACudaOnlyCUDA.test_flash_attention_vs_math_ref_grads_batch_size_8_seq_len_q_143_seq_len_k_4_head_dim_203_is_causal_False_dropout_p_0_22_float16_scale_l1_enable_gqa_True_n_heads1_cuda_float16`
, at least on ROCM platform. The false negative disappears after forcing
`higher_precision_dtype = torch.float64`

# Major Changes

To restore the precision of golden output, a wrapper function is used where golden solution is calculated.
This function sets FP32 precision in scope of calculation of golden solution and its gradient. 

This is based of PR https://github.com/pytorch/pytorch/pull/167157, changes are moved due to pytorchbot limited access to the ROCm fork.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @azahed98 @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @hongxiayang @naromero77amd @pragupta @jerrymannil

xinyazhang committed:
Add torch.backends.cuda.math_sdp.fp32_precision

xinyazhang committed:
Make torch.backends.cuda.math_sdp.fp32_precision effective for math_sdp

xinyazhang committed:
torch/testing: add ctx mananger math_sdp_precision

xinyazhang committed:
test/test_transformers: decorate all tests that uses fp32 math as golden to use ieee rather than tf32

xinyazhang committed:
fix build error

xinyazhang committed:
test/test_transformers: sanity check of golden tensor

xinyazhang committed:
fix Fp32PrecisonGuard

xinyazhang committed:
more documentation

xinyazhang committed:
fix lint

xinyazhang committed:
Follow the practice of cuBLASModule and ignore MathSDPModule in docs generation

anatoliylitv committed:
Follow the practice of public cuBLASModule to include to __all__  for  MathSDPModule

xinyazhang committed:
Revert "Follow the practice of cuBLASModule and ignore MathSDPModule in docs generation"

This reverts commit b996c2e2f44dbc4d952beb09bdb7dbfa84837d03.

Need to document this new API.

xinyazhang committed:
docs: document API torch.backends.cuda.math_sdp.fp32_precision

jeffdaily commented:
@pytorchbot merge -f "all CI passing except known mi355 issues"

drisspg commented:
Why dont the existing tf32 precision setter/getters handle this?

albanD commented:
@pytorchbot revert -m "As mentioned in my comment above, please wait for Driss to approve before merging" -c weird

drisspg commented:
why dont we have finer grained usage of the tf32 apis as opposed to adding new API surface to PT?

eqy reviewed (commented):
Pardon my naive understanding, but is this really just aliasing the matmul precision (as MATH calls into matmuls internally)? Why is a separate flag needed here instead of switching TF32 precision as needed in tests?

anatoliylitv committed:
Refactor TF32 context manager and update tests to use @tf32_enabled. Removed @math_sdp_precision in favor of new functionality in tf32_enabled to manage precision settings.

anatoliylitv committed:
Update tf32_enabled context manager parameters order which siplify calls from existing  test.

anatoliylitv committed:
Fix lint error.

anatoliylitv committed:
Fix trailing spaces lint issue.

anatoliylitv committed:
Refactored TF32 context manager to consolidate settings of multiple precision flags. The tf32_enabled decorator has been replaced w
ith a more versatile tf32 function, allowing for flexible control over precision  settings in tests.

anatoliylitv commented:
@pytorchbot -rebase

anatoliylitv commented:
@pytorchbot rebase

anatoliylitv committed:
Address new lint ban of use of asserts.

anatoliylitv commented:
> why dont we have finer grained usage of the tf32 apis as opposed to adding new API surface to PT?

Hi, @drisspg. 
The concerns you raised (as understood) was addressed by implementing finer‑grained control of the tf32 APIs rather than introducing additional precision settings decorators. The tf32 decorator now provides flexible management of multiple precision flags within tests, with a design that allows future expansion without requiring new public APIs.
I would appreciate any feedback you may have, and if possible, a review of the pull request.

drisspg reviewed (dismissed):
So your claim is that the golden reference is being calculated in to low of a precision because we are using tf32 for the matmuls and should be using fp32, instead of adding all this machinery should we not just update how the golden reference is testing to use fp64?

If for some reasons - maybe backend support or what have you we should still compute gold in fp32 I dont get why we cant use existing APIs to disable TF32 temporarily why calculating the gold reference intra test

drisspg commented:
@claude

Here was my comment "So your claim is that the golden reference is being calculated in to low of a precision because we are using tf32 for the matmuls and should be using fp32, instead of adding all this machinery should we not just update how the golden reference is testing to use fp64?

If for some reasons - maybe backend support or what have you we should still compute gold in fp32 I dont get why we cant use existing APIs to disable TF32 temporarily why calculating the gold reference intra test"

Can you verify this is possible without needing to add new finer grained controls to pytorch

anatoliylitv committed:
Revert "Merge sdp_ieee changes (#5)"

This reverts commit adde14f5c0940293b4bf2911201cd9e7bdbf2d40.

anatoliylitv committed:
Increase precision for computing golden solution in transformers tests.

anatoliylitv commented:
@pytorchbot rebase

anatoliylitv committed:
Add precision wrapper for golden reference computation in SDPA tests. (#6)

Add IEEE precision wrapper for golden reference computation in SDPA tests.

anatoliylitv committed:
Refactor SDPA tests to remove compute_golden_reference_ieee wrapper and directly use tf32_off() for high precision calculations

anatoliylitv committed:
Update SDPA tests to consistently apply tf32_off() for high precision calculations across multiple reference outputs

xinyazhang reviewed (commented):
The problem is you have to add ctx manager to all places you can find.
Consider adding helper functions for lp+golden tensor calculation as the uniform entrance to avoid missing things.

anatoliylitv commented:
@pytorchbot rebase

anatoliylitv commented:
@pytorchbot rebase

jansel reviewed (changes requested):
fix merge conflicts

anatoliylitv commented:
@pytorchbot rebase

anatoliylitv commented:
@pytorchbot rebase

jansel commented:
@drisspg should review this one

drisspg reviewed (commented):
d

jansel commented:
@drisspg

jansel reviewed (changes requested):
fix merge conflicts

anatoliylitv commented:
@pytorchbot rebase

jansel reviewed (changes requested):
fix merge conflicts

anatoliylitv committed:
Remove merge changes.

malfet reviewed (approved):
Hmm, is `tf32_off` specific to thread or applies to the entire stream?

anatoliylitv reviewed (commented):
On some future architectures fp64 may not be supported. Made changes to use existing APIs to disable TF32 temporarily why calculating the gold reference.

jeffdaily commented:
@pytorchbot merge -r
